sound/soc/sof/stream-ipc.c

```diff
@@ -36,20 +36,24 @@ int sof_ipc_msg_data(struct snd_sof_dev *sdev,
 		size_t posn_offset;
 
 		if (sps->substream) {
-			struct sof_stream *stream = sps->substream->runtime->private_data;
+			struct snd_pcm_runtime *runtime = sps->substream->runtime;
 
 			/* The stream might already be closed */
-			if (!stream)
+			if (!runtime || !runtime->private_data)
 				return -ESTRPIPE;
 
+			struct sof_stream *stream = runtime->private_data;
+
 			posn_offset = stream->posn_offset;
 		} else {
 
-			struct sof_compr_stream *sstream = sps->cstream->runtime->private_data;
+			struct snd_compr_runtime *runtime = sps->cstream->runtime;
 
-			if (!sstream)
+			if (!runtime || !runtime->private_data)
 				return -ESTRPIPE;
 
+			struct sof_compr_stream *sstream = runtime->private_data;
+
```

Comment:
I would probably add this as well:
```diff
diff --git a/sound/soc/sof/pcm.c b/sound/soc/sof/pcm.c
index 62ce707abaa6..7a55bc65bdf1 100644
--- a/sound/soc/sof/pcm.c
+++ b/sound/soc/sof/pcm.c
@@ -511,6 +511,8 @@ static int sof_pcm_close(struct snd_soc_component *component,
 		 */
 	}
 
+	spcm->stream[substream->stream].substream = NULL;
+
 	return 0;
 }
 ```
We never clear the substream, just set it... This alone might work?

Reply:
No this won't it. This would still result in the same bug because you would just then just hit a NULL pointer on the cstream.

Reply:
the sps->cstream is cleared to NULL in compress.c, only the PCM substream missed the reset to NULL.

Reply:
> the sps->cstream is cleared to NULL in compress.c, only the PCM substream missed the reset to NULL.

Right, you would hit different NULL as in `sof_ipc_msg_data()` the check for `sps->cstream == NULL` is missing and assumed to be non NULL if substream was NULL.

`sof_set_stream_data_offset()` does this correctly...

Reply:
this might work for the compress stream case:
```diff
diff --git a/sound/soc/sof/stream-ipc.c b/sound/soc/sof/stream-ipc.c
index 794c7bbccbaf..bcfb4114e4b6 100644
--- a/sound/soc/sof/stream-ipc.c
+++ b/sound/soc/sof/stream-ipc.c
@@ -43,7 +43,7 @@ int sof_ipc_msg_data(struct snd_sof_dev *sdev,
 				return -ESTRPIPE;
 
 			posn_offset = stream->posn_offset;
-		} else {
+		} else if (sps->cstream) {
 
 			struct sof_compr_stream *sstream = sps->cstream->runtime->private_data;
 
@@ -51,6 +51,9 @@ int sof_ipc_msg_data(struct snd_sof_dev *sdev,
 				return -ESTRPIPE;
 
 			posn_offset = sstream->posn_offset;
+		} else {
+			dev_err(sdev->dev, "%s: No stream opened\n", __func__);
+			return -EINVAL;
 		}
 
 		snd_sof_dsp_mailbox_read(sdev, posn_offset, p, sz);
```